src/clusterfuzz/_internal/bot/tasks/utasks/progression_task.py

```diff
@@ -403,83 +498,72 @@ def find_fixed_range(testcase_id, job_type):
     data_handler.update_testcase_comment(testcase, data_types.TaskState.WIP,
                                          log_message)
 
-    try:
-      result = _testcase_reproduces_in_revision(testcase, testcase_file_path,
-                                                job_type, middle_revision)
-    except errors.BadBuildError:
-      # Skip this revision.
-      del revision_list[middle_index]
-      max_index -= 1
-      continue
 
+    result, error = _testcase_reproduces_in_revision(testcase, testcase_file_path,
+                                                job_type, middle_revision)
+    if error is not None:
+      if error.error == uworker_msg_pb2.ErrorType.PROGRESSION_BAD_BUILD:
+        # Skip this revision.
+        del revision_list[middle_index]
+        max_index -= 1
+        continue
+      # only bad build errors are recoverable.
+      return error
+    
     if result.is_crash():
       min_index = middle_index
     else:
       max_index = middle_index
 
+    # TODO: this cannot be done from an untrusted bot.
     _save_current_fixed_range_indices(testcase_id, revision_list[min_index],
                                       revision_list[max_index])
 
   # If we've broken out of the loop, we've exceeded the deadline. Recreate the
   # task to pick up where we left off.
-  testcase = data_handler.get_testcase_by_id(testcase_id)
   error_message = ('Timed out, current range r%d:r%d' %
-                   (revision_list[min_index], revision_list[max_index]))
-  data_handler.update_testcase_comment(testcase, data_types.TaskState.ERROR,
-                                       error_message)
-  tasks.add_task('progression', testcase_id, job_type)
+                   (revision_list[min_index], revision_list[max_index]))  
+  return uworker_io.UworkerOutput(
+          testcase=testcase,
+          error_message=error_message,
+          uworker_input=uworker_input,
+          progression_min_revision=revision_list[min_index],
+          progression_max_revision=revision_list[max_index],
+          error=uworker_msg_pb2.ErrorType.PROGRESSION_TIMEOUT)
 
-
-def utask_preprocess(testcase_id, job_type, uworker_env):
-  return uworker_io.UworkerInput(
-      job_type=job_type,
-      testcase_id=testcase_id,
-      uworker_env=uworker_env,
-  )
+def utask_main(uworker_input):
+  """Execute progression task."""
+  return find_fixed_range(uworker_input)
 
 
 def utask_postprocess(output):
-  del output
-
+  if output.error is not None:
+    # TODO(https://github.com/google/clusterfuzz/issues/3008): Change this when
```

Comment:
This is the change. The TODO is completed and outdated now. Please remove it :-)